common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,1063 @@
+  // algorithm for sizes above 8, even if to just exercise the same code on
+  // more platforms.
+  if constexpr (Size > 8) {
+    static_assert(Size <= 32, "Sizes larger than 32 not yet supported!");
+    uint32_t bit = 1;
+    for (ssize_t i : llvm::seq<ssize_t>(0, Size)) {
+      if (metadata_bytes[i] == Deleted) {
+        return MatchIndex(bit);
+      }
+      bit <<= 1;
+    }
+    return MatchIndex(0);
+  }
+
+  // This sets the high bit of every byte in `match_bits` unless the
+  // corresponding metadata byte is 1. We take advantage of the fact that the
+  // metadata bytes are not 1 only if they are either:
+  // - present: in which case the high bit of the byte will already be set; or
+  // - empty: in which case the byte will be 0, and in that case inverting and
+  //   shifting left by 7 will have the high bit set.
+  uint64_t match_bits = metadata_ints[0] | (~metadata_ints[0] << 7);
+  // This inverts the high bits of the bytes, and clears the remaining bits.
+  match_bits = ~match_bits & MSBs;
+
+  // The high bits of the bytes of `match_bits` are set if the corresponding
+  // metadata byte is `Deleted`.
+  CARBON_DCHECK(VerifyIndexBits(match_bits,
+                                [](uint8_t byte) { return byte == Deleted; }));
+  return MatchIndex(match_bits);
+}
+
+inline auto MetadataGroup::PortableCompareEqual(MetadataGroup lhs,
+                                                MetadataGroup rhs) -> bool {
+  return llvm::equal(lhs.metadata_bytes, rhs.metadata_bytes);
+}
+
+inline auto MetadataGroup::SIMDLoad(const uint8_t* metadata, ssize_t index)
+    -> MetadataGroup {
+  MetadataGroup g;
+#if CARBON_NEON_SIMD_SUPPORT
+  g.metadata_vec = vld1_u8(metadata + index);
+#elif CARBON_X86_SIMD_SUPPORT
+  g.metadata_vec =
+      _mm_load_si128(reinterpret_cast<const __m128i*>(metadata + index));
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+  return g;
+}
+
+inline auto MetadataGroup::SIMDStore(uint8_t* metadata, ssize_t index) const
+    -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  vst1_u8(metadata + index, metadata_vec);
+#elif CARBON_X86_SIMD_SUPPORT
+  _mm_store_si128(reinterpret_cast<__m128i*>(metadata + index), metadata_vec);
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+}
+
+inline auto MetadataGroup::SIMDClearDeleted() -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  // There is no good Neon operation to implement this, so do it using integer
+  // code. This is reasonably fast, but unfortunate because it forces the group
+  // out of a SIMD register and into a general purpose register, which can have
+  // high latency.
+  metadata_ints[0] &= (~LSBs | metadata_ints[0] >> 7);
+#elif CARBON_X86_SIMD_SUPPORT
+  metadata_vec =
+      _mm_blendv_epi8(_mm_setzero_si128(), metadata_vec, metadata_vec);
```

Comment:
This is the level of detail I think would be appropriate to comment this code:
```suggestion
  // For each byte, use `metadata_vec` if the byte's high bit is set (indicating it is present), otherwise (it is empty or deleted) replace it with zero (representing empty).
  metadata_vec =
      _mm_blendv_epi8(_mm_setzero_si128(), metadata_vec, metadata_vec);
```